src/main/java/io/temporal/internal/statemachines/ActivityStateMachine.java

```diff
@@ -0,0 +1,404 @@
+    ScheduleActivityTaskCommandAttributes.Builder scheduleAttr = parameters.getAttributes();
+    ActivityFailureInfo failureInfo =
+        ActivityFailureInfo.newBuilder()
+            .setActivityId(scheduleAttr.getActivityId())
+            .setActivityType(scheduleAttr.getActivityType())
+            .setIdentity(failed.getIdentity())
+            .setRetryState(failed.getRetryState())
+            .setScheduledEventId(failed.getScheduledEventId())
+            .setStartedEventId(failed.getStartedEventId())
+            .build();
+    Failure failure =
+        Failure.newBuilder()
+            .setActivityFailureInfo(failureInfo)
+            .setCause(failed.getFailure())
+            .setMessage("Activity task failed")
+            .build();
+    completionCallback.apply(Optional.empty(), failure);
+  }
+
+  private void notifyTimedOut() {
+    ScheduleActivityTaskCommandAttributes.Builder scheduleAttr = parameters.getAttributes();
+
+    ActivityTaskTimedOutEventAttributes timedOut =
+        currentEvent.getActivityTaskTimedOutEventAttributes();
+
+    ActivityFailureInfo failureInfo =
+        ActivityFailureInfo.newBuilder()
+            .setActivityId(scheduleAttr.getActivityId())
+            .setActivityType(scheduleAttr.getActivityType())
+            .setRetryState(timedOut.getRetryState())
+            .setScheduledEventId(timedOut.getScheduledEventId())
+            .setStartedEventId(timedOut.getStartedEventId())
+            .build();
+    Failure failure =
+        Failure.newBuilder()
+            .setActivityFailureInfo(failureInfo)
+            .setCause(timedOut.getFailure())
+            .setMessage("Activity task timedOut")
+            .build();
+    completionCallback.apply(Optional.empty(), failure);
+  }
+
+  private void notifyCancellationFromEvent() {
+    if (parameters.getCancellationType() == ActivityCancellationType.WAIT_CANCELLATION_COMPLETED) {
+      ActivityTaskCanceledEventAttributes canceledAttr =
+          currentEvent.getActivityTaskCanceledEventAttributes();
+      Failure canceledFailure =
+          Failure.newBuilder()
+              .setSource(JAVA_SDK)
+              .setCanceledFailureInfo(
+                  CanceledFailureInfo.newBuilder().setDetails(canceledAttr.getDetails()))
+              .build();
+
+      ScheduleActivityTaskCommandAttributes.Builder scheduleAttr = parameters.getAttributes();
+      ActivityFailureInfo failureInfo =
+          ActivityFailureInfo.newBuilder()
+              .setActivityId(scheduleAttr.getActivityId())
+              .setActivityType(scheduleAttr.getActivityType())
+              .setScheduledEventId(canceledAttr.getScheduledEventId())
+              .setStartedEventId(canceledAttr.getStartedEventId())
+              .build();
+      Failure failure =
+          Failure.newBuilder()
+              .setActivityFailureInfo(failureInfo)
+              .setCause(canceledFailure)
+              .setMessage("Activity task timedOut")
```

Comment:
should this error message indicate the activity was canceled rather than timed out?

Reply:
Fixed